CMake: use python -m sphinx instead of sphinx-build et al.

mkuron opened the pull request:
This helps in situations where the python3 version of `sphinx-build` is not in $PATH, but the python2 version is.

Also apply the same trick to Cython, where we already used it as a fallback.

Problem reported by @biermanncarl and discussed with @KaiSzuttor.

mkuron committed:
CMake: use python -m sphinx instead of sphinx-build et al.

This helps in situations where the python3 version of sphinx-build is not in $PATH, but the python2 version is.

Also apply the same trick to Cython, where we already used it as a fallback.

jngrad commented on cmake/FindSphinx.cmake:
```diff
@@ -29,20 +17,19 @@ if(Sphinx_FIND_VERSION)
-  endif()
+set(SPHINX_VERSION_COMPATIBLE TRUE)
+# Blacklist broken version
+if("${SPHINX_VERSION}" VERSION_EQUAL "2.1.0")
+  message(WARNING "Sphinx version 2.1.0 is not compatible.")
```
to avoid conflicts with the Sphinx PR in #3642, we could either integrate this commit in #3642, or apply this patch:
```suggestion
if("${SPHINX_VERSION}" VERSION_EQUAL "2.1.0" OR "${SPHINX_VERSION}" VERSION_EQUAL "3.0.0")
  message(WARNING "Sphinx version ${SPHINX_VERSION} is not compatible.")
```

KaiSzuttor replied:
that is a trivial merge, right?

jngrad replied:
yes, but it's an annoying backport in 4.1.3 because merge commits break `git cherry-pick`

mkuron committed:
blacklist Sphinx 3.0.0

Co-Authored-By: Jean-Noël Grad <jgrad@icp.uni-stuttgart.de>

mkuron committed:
Fix sphinx version regex

some versions report __main__.py 2.4.4 instead of sphinx-build 2.4.4 and our regex matched on the first dot

jngrad commented on cmake/FindSphinx.cmake:
```diff
@@ -1,25 +1,13 @@
-                               $ENV{HOME}/Library/Python/2.7/ PATH_SUFFIXES bin
-  DOC "Sphinx documentation generator.")
+set(SPHINX_EXECUTABLE ${PYTHON_EXECUTABLE} -m sphinx)
+set(SPHINX_API_DOC_EXE ${PYTHON_EXECUTABLE} -m sphinx.apidoc)
```
This change and the one in the Cython file generate the following CMake log:
```
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.6.9", minimum required is "3.5") 
-- Found Cython: /usr/bin/python3;-m;cython (found suitable version "0.26.1", minimum required is "0.26") 
-- Found PythonHeaders: /usr/include/python3.6m  
-- Found NumPy: /tikhome/jgrad/.local/lib/python3.6/site-packages/numpy/core/include (found version "1.17.3") 
-- Found Sphinx: /usr/bin/python3;-m;sphinx (found suitable version "2.0.1", minimum required is "1.6.6") 
```
Is there a way to get the absolute path of Sphinx and Cython instead of a list cast to a string (`/usr/bin/python3;-m;sphinx`)?

jngrad replied:
if not, could we instead quote the value of `SPHINX_EXECUTABLE`, `SPHINX_API_DOC_EXE`, `CYTHON_EXECUTABLE` to prevent the cast to string that adds semicolons?

mkuron replied:
These semicolons are only added when printing the "Found XXX" message. The list is correctly unpacked when calling the program. If we convert the list to a string inside FindSphinx/FindCython, CMake later tries to call that whole string as the executable and obviously fails.

> Is there a way to get the absolute path of Sphinx and Cython

No. In fact, there might not even be one.

KaiSzuttor replied:
let's first get in functional and then pretty, CI is failing...

KaiSzuttor replied:
@jngrad what's the issue with the semicolons?

mkuron committed:
fix sphinx apidoc detection above version 1.7

mkuron commented:
Okay, seems to be working now.

KaiSzuttor commented:
I think this is also not needed... If you'd call the wrong executable on the command line because your ```PATH``` is not set correctly I do not expect Espresso to be able to guess where the "correct" executable can be found.

mkuron commented:
Sometimes there is no correct executable, so modifying `$PATH` gets you nowhere. See e.g. https://packages.ubuntu.com/bionic/all/python3-sphinx/filelist: it does not contain a `sphinx-build`.

KaiSzuttor commented:
it does, how should it work if now

KaiSzuttor commented:
second last line

mkuron commented:
> how should it work if not

It does not work on a freshly-installed Ubuntu 18.04. You cannot build documentation there without this patch. I tried it.

> second last line

Seriously, who adds `/usr/share/sphinx/scripts/python3` to their `$PATH` manually?

mkuron commented:
Also, this patch actually simplifies finding the tool because the process doesn't need to worry about paths.
